File: day17/.js
/**
 * @param {string[][]} board
 * @returns {boolean}
 */
function hasFourLights(board) {
  if (!board.length) return false

  const rows = board.length
  const cols = board[0].length

  function hasStreak(getCell, length){
    let count = 0
    let prev = '.'

    for (let i = 0; i < length; i++) {
      const cell = getCell(i)

      if (cell === '.') {
        count = 0
        prev = '.'
      } else if (cell === prev) {
        count++
      } else {
        prev = cell
        count = 1
      }

      if (count === 4) return true
    }

    return false
  }
  for (let r = 0; r < rows; r++) {
    if (hasStreak(c => board[r][c], cols)) {
      return true
    }
  }

  for (let c = 0; c < cols; c++) {
    if (hasStreak(r => board[r][c], rows)) {
      return true
    }
  }

  return false
}

File: day17/.py
def has_four_lights(board: list[list[str]]) -> bool:
  if not board:
    return False

  rows = len(board)
  cols = len(board[0])

  def has_streak(line: list[str]) -> bool:
    count = 0
    prev = "."

    for cell in line:
      if cell != "." and cell == prev:
        count += 1
      elif cell != ".":
        count = 1
        prev = cell
      else:
        count = 0
        prev = "."
      if count == 4:
        return True
    return False

    # rows
    for r in range(rows):
      if has_streak(board[r]):
        return True

    # cols
    for c in range(cols):
      col = [board[r][c] for r in range(rows)]
      if has_streak(col):
        return True

    return False
